Vokal-Swift.xctemplate/APICompletionTests.swift

```diff
@@ -0,0 +1,140 @@
+class APICompletionTests: XCTestCase {
+    
+    let StandardTestTimeout = 3.0
+    
+    //MARK: - Test Lifecycle
+    
+    override static func setUp() {
+        super.setUp()
+
+        //Use mock protocol
+        HTTPSessionManager.useMockData()
+    }
+
+    override static func tearDown() {
+        //Stop using mock protocol
+        HTTPSessionManager.useLiveData()
+
+        super.tearDown()
+    }
+
+    //MARK: - Test generic closures in API completion
+
+    func testDictionaryFetch() {
+        //GIVEN: Using mock data and an async method
+        let expectation = self.expectation(description: "Finish requesting dictionary")
+        
+        //WHEN: Requesting a URL that should return a dictionary
+        MainAPIUtility.sharedUtility.getJSON(from: "fetchDictionary",
+                                             headers: [:],
+                                             success: { (result: APIDictionary) in
+                                                //THEN: the result data type is dictionary as expected
+                                                expectation.fulfill()
+        }, failure: { error in
+            XCTFail("Failed to fetch a dictionary. Error: \(error)")
+            expectation.fulfill()
+        })
+        
+        self.waitForExpectations(timeout: StandardTestTimeout, handler: nil)
+    }
+
+    func testArrayFetch() {
+        //GIVEN: Using mock data and an async method
+        let expectation = self.expectation(description: "Finish requesting array")
+
+        //WHEN: Requesting a URL that should return an array
+        MainAPIUtility.sharedUtility.getJSON(from: "fetchArray",
+                                             headers: [:],
+                                             success: { (result: APIArray) in
+                                                //THEN: the result data type is array as expected
+                                                expectation.fulfill()
+        }, failure: { error in
+            XCTFail("Failed to fetch an array. Error: \(error)")
+            expectation.fulfill()
+        })
+        
+        self.waitForExpectations(timeout: StandardTestTimeout, handler: nil)
+    }
+
+    func testArrayFail() {
+        //GIVEN: Using mock data and an async method
+        let expectation = self.expectation(description: "Finish requesting array")
+
+        //WHEN: Requesting a URL that should return an array
+        MainAPIUtility.sharedUtility.getJSON(from: "failArray",
+                                             headers: [:],
+                                             success: { (result: APIArray) in
+                                                XCTFail("This should fail: mock data does not contain an array")
+                                                expectation.fulfill()
+        }, failure: { error in
+            //THEN: the result data type is NOT array as requested, becuase mock data contains a dictionary
+            switch error {
```

Comment:
`NetworkError` would need to be equatable for that to work. Because that enum has a couple cases with associated values (`undefinedError(statusCode: Int)`, `otherError(error: Error)`), that comparison will get a lot uglier than this switch—especially since `Error` isn't equatable.